native/cocos/core/memop/CachedArray.h

```diff
@@ -1,192 +1,167 @@
 
 namespace cc {
 
-/**
- * @en
- * Cached array is a data structure for objects cache, it's designed for persistent data.
- * Its content array length will keep grow.
- * @zh
- * 适用于对象缓存的数组类型封装，一般用于不易被移除的常驻数据。
- * 它的内部数组长度会持续增长，不会减少。
- */
 template <typename T>
 class CachedArray final {
 public:
-    using CompareFn = std::function<int32_t(T *, T *)>;
+    explicit CachedArray(uint size = 1U) {
+        _size = 0;
+        _capacity = std::max(size, 1U);
+        _array = ccnew T[_capacity];
+    }
 
-    explicit CachedArray(uint32_t length) {
-        _array.reverse(length);
+    // The rule of five applies here
+    ~CachedArray() {
+        CC_SAFE_DELETE_ARRAY(_array);
     }
 
-    /**
-     * @param length Initial length
-     * @param compareFn Comparison function for sorting
-     */
-    CachedArray(uint32_t length, const CompareFn &compareFn) {
-        _array.reverse(length);
-        _compareFn = compareFn;
+    CachedArray(const CachedArray &other)
+    : _size(other._size), _capacity(other._capacity), _array(ccnew T[other._capacity]) {
+        memcpy(_array, other._array, _size * sizeof(T));
     }
 
-    CachedArray(const CachedArray &) = delete;
-    CachedArray(CachedArray &&) = delete;
-    CachedArray &operator=(const CachedArray &) = delete;
-    CachedArray &operator=(CachedArray &&) = delete;
-
-    /**
-     * @en
-     * Push an element to the end of the array
-     * @zh
-     * 向数组末尾添加一个元素
-     * @param item The item to be added
-     */
-    void push(T *item) {
-        if (_length < _array.capacity()) {
-            _array[_length++] = item;
+    CachedArray &operator=(const CachedArray &other) {
+        if (this != &other) {
+            delete[] _array;
+            _size = other._size;
+            _capacity = other._capacity;
+            _array = ccnew T[_capacity];
+            memcpy(_array, other._array, _size * sizeof(T));
         }
+        return *this;
     }
 
-    /**
-     * @en
-     * Pop the last element in the array. The [[length]] will reduce, but the internal array will keep its size.
-     * @zh
-     * 弹出数组最后一个元素，CachedArray 的 [[length]] 会减少，但是内部数组的实际长度不变
-     * @return The last element.
-     */
-    T *pop() {
-        if (_length == 0) {
-            return nullptr;
-        }
-        return _array[_length--];
+    CachedArray(CachedArray &&other) noexcept : _size(other._size), _capacity(other._capacity), _array(other._array) {
+        other._size = 0;
+        other._capacity = 0;
+        other._array = nullptr;
     }
 
-    /**
-     * @en
-     * Get the element at the specified index of the array
-     * @zh
-     * 得到数组中指定位置的元素
-     * @param idx The index of the requested element
-     * @return The element at given index
-     */
-    T *get(uint32_t idx) const {
-        if (idx < _length) {
-            return _array[idx];
+    CachedArray &operator=(CachedArray &&other) noexcept {
+        if (this != &other) {
+            delete[] _array;
```

Comment:
Use CC_SAFE_DELETE_ARRAY instead of delete[] ?

Reply:
Can do in other PR.